localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py

```diff
@@ -81,101 +104,129 @@ def poll_events(self):
-                    return
-                except PartialBatchFailureError:
-                    # TODO: add tests for partial batch failure scenarios
-                    if (
-                        self.stream_parameters["OnPartialBatchItemFailure"]
-                        == OnPartialBatchItemFailureStreams.AUTOMATIC_BISECT
-                    ):
-                        # TODO: implement and test splitting batches in half until batch size 1
-                        #  https://docs.aws.amazon.com/eventbridge/latest/pipes-reference/API_PipeSourceKinesisStreamParameters.html
-                        LOG.warning(
-                            "AUTOMATIC_BISECT upon partial batch item failure is not yet implemented. Retrying the entire batch."
-                        )
-
-                    # let entire batch fail (ideally raise BatchFailureError)
-                    LOG.debug(
-                        f"Attempt {attempts} failed while processing {self.partner_resource_arn} with events: {events}"
-                    )
-                    attempts += 1
-                    # Retry polling until the record expires at the source
-                    if self.stream_parameters.get("MaximumRetryAttempts", -1) == -1:
-                        return
-                except Exception:
-                    LOG.debug(
-                        f"Attempt {attempts} failed while processing {self.partner_resource_arn} with events: {events}"
-                    )
-                    attempts += 1
-                    # Retry polling until the record expires at the source
-                    if self.stream_parameters.get("MaximumRetryAttempts", -1) == -1:
-                        return
-
-            # Send failed events to potential DLQ
-            abort_condition = abort_condition or "RetryAttemptsExhausted"
-            context = {
-                "condition": abort_condition,
-                "partnerResourceArn": self.partner_resource_arn,
-            }
-            self.send_events_to_dlq(events, context=context)
-            # Update shard iterator upon failure once the events are sent to the DLQ
-            self.shards[shard_id] = get_records_response["NextShardIterator"]
-        else:
-            # Set current shard iterator to None to re-start round-robin at the first shard
-            self.iterator_over_shards = None
-
-    @abstractmethod
-    def initialize_shards(self) -> dict[str, str]:
-        """Returns a shard dict mapping from shard id -> shard iterator
-        The implementations for Kinesis and DynamoDB are similar but differ in various ways:
-        * Kinesis uses "StreamARN" and DynamoDB uses "StreamArn" as source parameter
-        * Kinesis uses "StreamStatus.ACTIVE" and DynamoDB uses "StreamStatus.ENABLED"
-        * Only Kinesis supports the additional StartingPosition called "AT_TIMESTAMP" using "StartingPositionTimestamp"
-        """
-        pass
+            self.shards = self.initialize_shards()
```

Comment:
As we discussed, we should really keep an eye on this and do more testing, as this might start the stream from the beginning if `TRIM_HORIZON` or `AT_...` is the iterator type.

Reply:
agree, I added an explicit TODO clarifying the issue here 👍